broker/request_manager.go

```diff
@@ -0,0 +1,106 @@
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package broker
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type (
+	reversalRequest struct {
+		CallbackUrl string `json:"callback_url"`
+		PrivateKey  string `json:"private_key"`
+		RequestId   string `json:"request_id"`
+		Prefix      string `json:"prefix"`
+		OriginName  string `json:"origin"`
+	}
+
+	requestInfo struct {
+		channel chan reversalRequest
+		prefix  string
+	}
+
+	requestKey struct {
+		origin string
+		prefix string
+	}
+)
+
+var (
+	errRetrieveTimeout error                      = errors.New("retrieve request timed out")
+	errRequestTimeout  error                      = errors.New("reverse request timed out")
+	requestsLock       sync.Mutex                 = sync.Mutex{}
+	requests           map[requestKey]requestInfo = make(map[requestKey]requestInfo)
+)
+
+func getOriginQueue(prefix, origin string) chan reversalRequest {
+	requestsLock.Lock()
+	defer requestsLock.Unlock()
+	if req, ok := requests[requestKey{origin: origin, prefix: prefix}]; ok {
+		return req.channel
+	} else {
+		newChan := make(chan reversalRequest)
+		requests[requestKey{origin: origin, prefix: prefix}] = requestInfo{channel: newChan, prefix: prefix}
+		return newChan
+	}
+}
+
+// Send a request to a given origin's queue.
+// Return a requestTimeout error if no origin retrieved the request before the context timed out.
+func handleRequest(ctx context.Context, origin string, req reversalRequest, timeout time.Duration) (err error) {
+	queue := getOriginQueue(req.Prefix, origin)
+	maxTime := timeout - 500*time.Millisecond - time.Duration(rand.Intn(500))*time.Millisecond
```

Comment:
`timeout` was not bounded by any lower number and it can be less than 500ms, which will cause `NewTicker` to panic.

Reply:
Fixed.